# List of Contents

### 01 Importing libraries and data file Q1

### 02 Grouping and Aggregating

### 03 Exporting in pkl format Q9

# 01 Importing libraries and data file Q1

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Assign a path
path = r'/Users/dramblys/Desktop/Achievement 4/Instacart Basket Analysis'

In [3]:
# Import orders_products_merged_new.pkl
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new.pkl'))

In [4]:
# Check shape
ords_prods_merge.shape

(32404859, 21)

# 02 Grouping and Aggregating

In [5]:
# Find mean of 'order_number', grouped by 'department-id' Q2
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

Q3: The sample of the first 1 million rows is a poor representation of the entire dataset. The sample did not include all departments (only 8/21) and the means for departments that were included differ quite significantly from those for the entire dataset. For example, sample mean for department 17 was 11.29, where for the entire set it is 15.69.

In [6]:
# Create 'max_order' column with transform Q4
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
# Create loyalty flag with loc Q4
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [8]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [9]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Check counts
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [11]:
# Group by Loyalty and get basic stats Q5
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['min', 'max','mean','median']})

prices                           
                    min      max       mean median
loyalty_flag                                      
Loyal customer      1.0  99999.0  10.386336    7.4
New customer        1.0  99999.0  13.294670    7.4
Regular customer    1.0  99999.0  12.495717    7.4

In [13]:
# Check for unreasonable high prices
ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

Unnamed: 0_x  product_id                      product_name  \
13100147         21554       21553  Lowfat 2% Milkfat Cottage Cheese   
13100148         21554       21553  Lowfat 2% Milkfat Cottage Cheese   
13100149         21554       21553  Lowfat 2% Milkfat Cottage Cheese   
13100150         21554       21553  Lowfat 2% Milkfat Cottage Cheese   
13100151         21554       21553  Lowfat 2% Milkfat Cottage Cheese   
...                ...         ...                               ...   
21786876         33666       33664             2 % Reduced Fat  Milk   
21786877         33666       33664             2 % Reduced Fat  Milk   
21786878         33666       33664             2 % Reduced Fat  Milk   
21786879         33666       33664             2 % Reduced Fat  Milk   
21786880         33666       33664             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices  Unnamed: 0.1  Unnamed: 0_y  \
13100147       108             16  14900.0           171           171   
13100148       108             16  14900.0           181           181   
13100149       108             16  14900.0          1944          1944   
13100150       108             16  14900.0          1945          1945   
13100151       108             16  14900.0          5708          5708   
...            ...            ...      ...           ...           ...   
21786876        84             16  99999.0       3386542       3386542   
21786877        84             16  99999.0       3386544       3386544   
21786878        84             16  99999.0       3391313       3391313   
21786879        84             16  99999.0       3405096       3405096   
21786880        84             16  99999.0       3405099       3405099   

          order_id  user_id  ...  days_since_prior_order  add_to_cart_order  \
13100147    912404       17  ...                     5.0                  5   
13100148    603376       17  ...                     4.0                  3   
13100149   3264360      135  ...                    13.0                  6   
13100150    892534      135  ...                    12.0                  3   
13100151    229704      342  ...                    30.0                  9   
...            ...      ...  ...                     ...                ...   
21786876   2249946   204099  ...                     4.0                  1   
21786877   2363282   204099  ...                     2.0                  1   
21786878   3181945   204395  ...                     8.0                 25   
21786879   2486215   205227  ...                     4.0                  8   
21786880   2232716   205227  ...                     4.0                 23   

          reordered  _merge     price_range_loc     busiest_day  \
13100147          0    both  High-range product  Regularly busy   
13100148          1    both  High-range product  Regularly busy   
13100149          0    both  High-range product  Regularly busy   
13100150          1    both  High-range product     Busiest day   
13100151          0    both  High-range product  Regularly busy   
...             ...     ...                 ...             ...   
21786876          0    both  High-range product     Busiest day   
21786877          1    both  High-range product     Busiest day   
21786878          0    both  High-range product  Regularly busy   
21786879          0    both  High-range product  Regularly busy   
21786880          1    both  High-range product     Busiest day   

            busiest_days busiest_period_of_day max_order      loyalty_flag  
13100147  Regularly busy           Most orders        40  Regular customer  
13100148  Regularly busy           Most orders        40  Regular customer  
13100149  Regularly busy        Average orders         4      New customer  
13100150    Busiest days        Average orders         4      New customer  
13100151    Busiest days        Average orders        16  Regular customer  
...                  ...                

In [14]:
# Turn "unreasonable" prices into NaNs
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [15]:
# Create 'average_price' column with transform Q6
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [16]:
# Group by Loyalty and get basic stats Q5
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['min', 'max','mean','median']})

prices                       
                    min   max      mean median
loyalty_flag                                  
Loyal customer      1.0  25.0  7.773575    7.4
New customer        1.0  25.0  7.801206    7.4
Regular customer    1.0  25.0  7.798262    7.4

In [18]:
# Create spending flag with loc Q6
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [17]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [19]:
# Check counts
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32285131
High spender      119728
Name: spending_flag, dtype: int64

In [20]:
# Create 'median_order_frequency' column with transform Q7
ords_prods_merge['median_order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [21]:
# Create order frequency flag with loc Q7
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [22]:
ords_prods_merge.loc[(ords_prods_merge['median_order_frequency'] > 10) & (ords_prods_merge['median_order_frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [23]:
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [24]:
# Check counts
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

# 03 Exporting in pkl format Q9

In [25]:
# Export in pickle format Q9
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_very_new.pkl'))